### PR TITLE
Client table API and client-side implementation.

### PR DESCRIPTION
**Pull request by petermattis**
Added {Get,Put,Inc,Scan,Del}Struct() which perform operations on structured tables. The initial implementation is 100% client side.

### CODE REVIEW

**Comment by petermattis** [comment]
As a first pass, I'd recommend against looking at the implementation in `table.go` and simply looking at the test in `table_test.go`. Yes, I realize this has inadequate testing.

**Comment by petermattis** [comment]
Btw, when you run the new `TestStruct` test with verbose logging you can see the underlying keys being accessed:

```
I0611 14:39:32.149006   55874 table.go:405] Put "users\x00\x01\t\x01title" ->
I0611 14:39:32.149028   55874 table.go:405] Put "users\x00\x01\t\x01delinquent" -> 0
I0611 14:39:32.149037   55874 table.go:405] Put "users\x00\x01\t\x01name" -> Peter
I0611 14:39:32.174736   55874 table.go:405] Put "users\x00\x01\t\x02title" -> CEO
I0611 14:39:32.174754   55874 table.go:405] Put "users\x00\x01\t\x02delinquent" -> 0
I0611 14:39:32.174764   55874 table.go:405] Put "users\x00\x01\t\x02name" -> Spencer
I0611 14:39:32.176630   55874 table.go:358] Get "users\x00\x01\t\x01name"
I0611 14:39:32.177489   55874 table.go:358] Get "users\x00\x01\t\x02title"
I0611 14:39:32.177500   55874 table.go:358] Get "users\x00\x01\t\x02delinquent"
I0611 14:39:32.177507   55874 table.go:358] Get "users\x00\x01\t\x02name"
I0611 14:39:32.178415   55874 table.go:625] Del "users\x00\x01\t\x02title"
I0611 14:39:32.178982   55874 table.go:358] Get "users\x00\x01\t\x02title"
I0611 14:39:32.178993   55874 table.go:358] Get "users\x00\x01\t\x02delinquent"
I0611 14:39:32.182978   55874 table.go:358] Get "users\x00\x01\t\x02name"
I0611 14:39:32.184623   55874 table.go:457] Inc "users\x00\x01\t\x02delinquent"
I0611 14:39:32.185348   55874 table.go:524] Scan "users\x00\x01\b" "users\x00\x01\n\x03\xe8"
I0611 14:39:32.185899   55874 table.go:524] Scan "users\x00\x01\b" "users\x00\x01\n\x03\xe8"
```

File: client/reflect_test.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
+// Copyright 2014 Square Inc.
```
Comment (petermattis) [inline client/reflect_test.go]: Taken from `github.com/square/squalor`.

File: client/reflect.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
+// Copyright 2014 Square Inc.
```
Comment (petermattis) [inline client/reflect.go]: Taken from `github.com/square/squalor`.

**Comment by tschottdorf** [comment]
The handling of this feels nice. Are there going to be situations in which we're trying to use structured data but we don't have struct definitions that basically match up with the schema? Where you may simply be trying to operate on tables and columns by their names? For example, a tool that gets a table and primary key column as strings and gives you the complete first row.

**Comment by petermattis** [comment]
Yes, I imagine we will have situations where we don't have struct definitions but still want the whole row. I was planning to add a mode where you could use a `map[string]interface{}` in such situations. The idea is that a struct definition is essentially a typed and restricted map.

**Comment by vivekmenezes** [comment]
Pete, do you want us to review this?

**Comment by petermattis** [comment]
@vivekmenezes Yes, this needs to be reviewed. My original comment still stands, though. Take a look at `table_test.go` so we can pick apart the API first before dissecting the implementation in `table.go`.

File: client/table_test.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	u = &User{ID: 2}
+	if err := db.GetStruct(u); err != nil {
+		t.Fatal(err)
+	}
+	if !reflect.DeepEqual(spencer, *u) {
+		t.Errorf("expected '%+v', but got '%+v'", spencer, *u)
+	}
+
+	// Verify we can delete a single column.
+	if err := db.DelStruct(spencer, "title"); err != nil {
```
Comment (tamird) [inline client/table_test.go]: should `DelStruct` update the struct?

File: client/table.go [hunk 4]
```diff
@@ -0,0 +1,781 @@
...
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return key, nil
+}
+
+// encodeColumnKey encodes the column and appends it to primaryKey.
+func (m *model) encodeColumnKey(primaryKey []byte, column string) []byte {
```
Comment (tamird) [inline client/table.go]: why deal in `[]byte` rather than `Key`?

File: client/table.go [hunk 5]
```diff
@@ -0,0 +1,781 @@
...
+	fields, err := getDBFields(t)
+	if err != nil {
+		return err
+	}
+	m := &model{
+		name:       name,
+		fields:     fields,
+		primaryKey: primaryKey,
+	}
+	isPrimaryKey := make(map[string]bool)
```
Comment (tamird) [inline client/table.go]: looks like a set

**Comment by tamird** [comment]
LGTM but the amount of `interface{}` magic here is terrifying

File: client/table.go (hunk 4 above)
Comment (petermattis) [inline client/table.go]: Well, the encoding routines take `[]byte` and not `Key`. That's not a very strong reason though. I'll change this to `Key`.

File: client/table.go (hunk 5 above)
Comment (petermattis) [inline client/table.go]: It is a set. Not quite clear what the hangup is about using `map[string]bool` in this manner. For temporary maps like this it seems quite natural.

File: client/table_test.go (hunk 3 above)
Comment (petermattis) [inline client/table_test.go]: Hmm, that's a good point. I can see arguments on both sides. Let me add a TODO to the code.

File: client/table.go (hunk 4 above)
Comment (petermattis) [inline client/table.go]: Hmm, the code started getting a lot more casts when trying to use `proto.Key` here. I'm going to leave as-is for now.

File: client/table.go (hunk 5 above)
Comment (tamird) [inline client/table.go]: it can be misleading

Comment (petermattis) [inline client/table.go]: Is it misleading in this usage?

Comment (tamird) [inline client/table.go]: it was misleading enough that i thought "where is he setting it to false?"

Comment (petermattis) [inline client/table.go]: Fair enough. Changed.

**Comment by petermattis** [comment]
@tamird The `interface{}` magic is pretty much required by Go for using structs in this manner. See `json.Marshal` for an example in the standard library that is conceptually similar (loops over the fields in a struct) and takes an `interface{}` parameter.

**Comment by tamird** [comment]
@petermattis understood that there's no alternative, pretty unfortunate

**Pull request merged**